### PR TITLE
feat: base addition for context handling

### PR DESCRIPTION
**Pull request by barelyhuman**
https://github.com/pmndrs/valtio/issues/204#issuecomment-890981138

Note: Needs additional test cases and checks , this is just a base commit for the feature, not the completed version.

### CODE REVIEW

**Comment by barelyhuman** [comment]
@dai-shi  I have a version that detects this 
```js
// invalid
const initialObj = {
  count: 0,
  inc() {
    ++this.count
  },
}

const state = proxy(initialObj)
```
though it's the hackiest solution I've ever written, so I'm testing it with other common setups so it doesn't collide with them. 
I'll let you know either way.

**Comment by barelyhuman** [comment]
@dai-shi want me to set it as ready for review or is there another case we need to check, let me know

**Comment by dai-shi** [comment]
If you don't have any concerns, it means ready!

**Comment by barelyhuman** [comment]
Where the warning will be shown in both cases.

Screenshot 2021-09-04 at 11 20 31 AM

File: src/AvoidThisInProxy.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import {
+  getParentOfNodeType,
+  isInSomething,
+  nearestCalleeName,
+} from './lib/utils'
+
+export const MESSAGE_THIS_IN_PROXY = `Avoid using \`this\` in valtio.proxy context.It might lead to unexpected results.
+Using this is valid, but often a pitfall for beginners.`
```
Comment (dai-shi) [inline src/AvoidThisInProxy.js:8]:
```suggestion
Using \`this\` is valid, but often a pitfall for beginners.`
```
Does this work?

**Review by dai-shi — APPROVED** [review]
LGTM. Nice work.

File: src/AvoidThisInProxy.js (hunk 1 above)
Comment (barelyhuman) [reply src/AvoidThisInProxy.js:8]: why I did i not see that....
Yeah works!

**Review by Aslemammad — COMMENTED** [review]
That's a clean & valid code! Thank you for your efforts.

**Pull request merged**